subscription_oca/models/sale_subscription.py

```diff
@@ -0,0 +1,807 @@
+# Copyright 2023 Domatix - Carlos Martínez
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+import logging
+from datetime import date, datetime
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import _, api, fields, models
+from odoo.exceptions import AccessError
+from odoo.tools.misc import get_lang
+
+logger = logging.getLogger(__name__)
+ODOO_DATE_FORMAT = "%Y-%m-%d"
+
+
+class SaleSubscription(models.Model):
+    _name = "sale.subscription"
+    _description = "Subscription"
+    _inherit = ["mail.thread", "mail.activity.mixin"]
+    _order = "id desc"
+
+    color = fields.Integer("Color Index")
+
+    def cron_subscription_management(self):
+        today = date.today()
+        query = []
+        for subscription in self.search(query):
+            if subscription.in_progress:
+                if (
+                    subscription.recurring_next_date == today
+                    and subscription.sale_subscription_line_ids
+                ):
+                    try:
+                        subscription.generate_invoice()
+                        subscription._onchange_template_id()
+                    except Exception:
+                        logger.exception("Error during Subscriptions Management cron")
+                if not subscription.recurring_rule_boundary:
+                    if subscription.date == today:
+                        subscription.action_close_subscription()
+
+            else:
+                if subscription.date_start == today:
+                    subscription.action_start_subscription()
+                    subscription.generate_invoice()
+                    subscription._onchange_template_id()
+
+    @api.depends("sale_subscription_line_ids")
+    def _compute_total(self):
+        for record in self:
+            recurring_total = amount_tax = 0.0
+            for order_line in record.sale_subscription_line_ids:
+                recurring_total += order_line.price_subtotal
+                amount_tax += order_line.amount_tax_line_amount
+            record.update(
+                {
+                    "recurring_total": recurring_total,
+                    "amount_tax": amount_tax,
+                    "amount_total": recurring_total + amount_tax,
+                }
+            )
+
+    def _compute_name(self):
+        template_code = self.template_id.code if self.template_id.code else ""
+        code = "{}".format(self.code) if self.code else ""
+        slash = "/" if template_code and code else ""
+        self.name = "{}{}{}".format(template_code, slash, code)
+
+    @api.onchange("template_id", "date_start")
+    def _onchange_template_id(self):
```

Comment:
this onchange seems to carry some business logic. I'd strongly recommend to use a computed field instead (then you should be able to avoid calling it everywhere)

Reply:
Any answer to this? If you cannot address this now, please add a TODO

Reply:
There was more functionality in the initial version, but I couldn't completely remove this without breaking the flow. Can you tell me how to fix it or what exactly I have to put in the TODO?

Reply:
Something like this would work. 
```
# TODO: refactor all the onchanges that have business logic
# to computed write-able fields when possible.
# Keep onchanges only for UI purposes
```
You can add this to the roadmap as well so that is prominent in the README.
And since you are there, you could add a big TODO for writing tests :wink: